### PR TITLE
Remove some usages of Guava in API

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/debt/DebtRemediationFunction.java [hunk 1]
```diff
@@ -85,25 +80,32 @@ public boolean equals(Object o) {
...
     DebtRemediationFunction that = (DebtRemediationFunction) o;
-    return new EqualsBuilder()
-      .append(type, that.type())
-      .append(coefficient, that.coefficient())
-      .append(offset, that.offset())
-      .isEquals();
+    if (type != that.type) {
+      return false;
+    }
+    if (coefficient != null ? !coefficient.equals(that.coefficient) : that.coefficient != null) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/debt/DebtRemediationFunction.java]: MAJOR Add parentheses to make the operator precedence explicit. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS864)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/debt/DebtRemediationFunction.java [hunk 2]
```diff
@@ -85,25 +80,32 @@ public boolean equals(Object o) {
...
+      return false;
+    }
+    return offset != null ? offset.equals(that.offset) : that.offset == null;
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/debt/DebtRemediationFunction.java]: MAJOR Add parentheses to make the operator precedence explicit. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS864)

File: sonar-plugin-api/src/main/java/org/sonar/api/issue/Issue.java [hunk 3]
```diff
@@ -67,14 +68,14 @@
    */
   String RESOLUTION_WONT_FIX = "WONTFIX";
 
-  List<String> RESOLUTIONS = ImmutableList.of(RESOLUTION_FALSE_POSITIVE, RESOLUTION_WONT_FIX, RESOLUTION_FIXED, RESOLUTION_REMOVED);
+  List<String> RESOLUTIONS = asList(RESOLUTION_FALSE_POSITIVE, RESOLUTION_WONT_FIX, RESOLUTION_FIXED, RESOLUTION_REMOVED);
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/issue/Issue.java]: MINOR Move "RESOLUTIONS" to a class and lower its visibility [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2386)

File: sonar-plugin-api/src/main/java/org/sonar/api/issue/Issue.java [hunk 4]
```diff
@@ -67,14 +68,14 @@
...
 
   /**
    * Return all available statuses
    *
    * @since 4.4
    */
-  List<String> STATUSES = ImmutableList.of(STATUS_OPEN, STATUS_CONFIRMED, STATUS_REOPENED, STATUS_RESOLVED, STATUS_CLOSED);
+  List<String> STATUSES = asList(STATUS_OPEN, STATUS_CONFIRMED, STATUS_REOPENED, STATUS_RESOLVED, STATUS_CLOSED);
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/issue/Issue.java]: MINOR Move "STATUSES" to a class and lower its visibility [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2386)

File: sonar-plugin-api/src/main/java/org/sonar/api/rule/Severity.java [hunk 5]
```diff
@@ -37,7 +37,7 @@
   /**
    * All the supported severity values, ordered from {@link #INFO} to {@link #BLOCKER}.
    */
-  public static final List<String> ALL = ImmutableList.of(INFO, MINOR, MAJOR, CRITICAL, BLOCKER);
+  public static final List<String> ALL = asList(INFO, MINOR, MAJOR, CRITICAL, BLOCKER);
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/rule/Severity.java]: MINOR Make this member "protected". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2386)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/coverage/CoverageType.java [hunk 6]
```diff
@@ -77,7 +79,8 @@
     this.uncoveredConditions = uncoveredConditions;
     this.coveredConditionsByLine = coveredConditionsByLine;
     this.conditionsByLine = conditionsByLine;
-    this.all = ImmutableSet.of(linesToCover, uncoveredLines, lineHitsData, conditionsToCover, uncoveredConditions, coveredConditionsByLine, conditionsByLine);
+    this.all = Collections
+      .unmodifiableSet(new LinkedHashSet<>(Arrays.asList(linesToCover, uncoveredLines, lineHitsData, conditionsToCover, uncoveredConditions, coveredConditionsByLine, conditionsByLine)));
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/coverage/CoverageType.java]: MINOR Split this 186 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: sonar-plugin-api/src/main/java/org/sonar/api/issue/DefaultTransitions.java [hunk 7]
```diff
@@ -43,5 +43,5 @@
   /**
    * @since 4.4
    */
-  List<String> ALL = ImmutableList.of(CONFIRM, UNCONFIRM, REOPEN, RESOLVE, FALSE_POSITIVE, WONT_FIX, CLOSE);
+  List<String> ALL = asList(CONFIRM, UNCONFIRM, REOPEN, RESOLVE, FALSE_POSITIVE, WONT_FIX, CLOSE);
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/issue/DefaultTransitions.java]: MINOR Move "ALL" to a class and lower its visibility [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2386)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [RulesProfile.java#L362](https://github.com/SonarSource/sonarqube/blob/37772bdf8ad6ca7882514800b93a7994269d346c/sonar-plugin-api/src/main/java/org/sonar/api/profiles/RulesProfile.java#L362): Remove this unused private "MatchRule" class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3985)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/DefaultFileSystem.java [hunk 8]
```diff
@@ -175,7 +173,9 @@ public boolean hasFiles(FilePredicate predicate) {
   @Override
   public Iterable<File> files(FilePredicate predicate) {
     doPreloadFiles();
-    return Iterables.transform(inputFiles(predicate), InputFile::file);
+    return () -> StreamSupport.stream(inputFiles(predicate).spliterator(), false)
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/DefaultFileSystem.java]: Do you really need "() -> " ?

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/cpd/internal/DefaultCpdTokens.java [hunk 9]
```diff
@@ -87,7 +88,7 @@ public DefaultCpdTokens addToken(TextRange range, String image) {
     if (excluded) {
       return this;
     }
-    Preconditions.checkState(lastRange == null || lastRange.end().compareTo(range.start()) <= 0,
+    checkState(lastRange == null || lastRange.end().compareTo(range.start()) <= 0,
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/cpd/internal/DefaultCpdTokens.java]: If the goal is to remove Guava, you should replace checkState by a if(...) throw ISE

Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/cpd/internal/DefaultCpdTokens.java]: This is my next proposal, that's why I didn't touch any of these Preconditions uses ! I suggest to create `org.sonar.api.utils.MoreUtils` with simple methods like checkState(), checkArgument() or isEmpty(), like what was done here : https://github.com/SonarSource/orchestrator/blob/master/sonar-orchestrator/src/main/java/com/sonar/orchestrator/util/OrchestratorUtils.java

Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/cpd/internal/DefaultCpdTokens.java]: Perfect, it sounds very good !

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/DefaultFileSystem.java (hunk 8 above)
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/DefaultFileSystem.java]: yes, that's the way to convert an Iterator to an Iterable.

Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/DefaultFileSystem.java]: ok !

**Pull request merged**